bin/gaia_marionette_retry.js

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env node
+var Promise = require('es6-promise').Promise;
+var format = require('util').format;
+var path = require('path');
+var spawn = require('child_process').spawn;
+
+function main() {
+  // 0 => node
+  // 1 => ./bin/gaia_marionette_retry.js
+  process.argv.splice(0, 2);
+
+  var index = 0;
+  while (process.argv[index].slice(0, 2) === '--') {
+    index += 2;
+  }
+
+  var args = process.argv.splice(0, index);
+  var filenames = process.argv;
+
+  // Also look at the end of the cli invocation for cli args.
+  var count = 0;
+  index = filenames.length;
+  while (filenames[index - 2].slice(0, 2) === '--') {
+    count += 2;
+    index -= 2;
+  }
+
+  args = args.concat(filenames.splice(index, count));
+
+  if (filenames[0].indexOf('setup') !== -1) {
+    // Special case for setup script.
+    args.push(filenames.shift());
+  }
+
+  runTests(filenames, args, 5 /* retry */).then(function(results) {
+    // Summarize.
+    console.log('*~*~* Results *~*~*');
+    console.log('passed: %d', results.pass);
+    console.log('failed: %d', results.fail);
+    console.log('todo: %d', results.pending);
+  });
+}
+
+function runTests(filenames, args, retry) {
```

Comment:
`runTests` returns a promise with the tallied result of running all the files in the filenames array. Internally, it uses recursion where the base case is no more files and the inductive case runs one test file (perhaps with retries) and augments the running results with an incremental result from the one test file.